File: Data Structures & Algorithms/anagram-groups/submission-2.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        unordered_map<string, vector<string>> hm;
        for(string& str : strs){
            string key = str;
            sort(key.begin(), key.end());
            hm[key].push_back(str);
        }

        for(auto& [k, v] : hm){
            res.push_back(v);
        }

        return res;

    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # bucket sort, assign them to buckets of frequencies
        n = len(nums)
        bucket = [[] for i in range(n+1)]

        count = Counter(nums)

        for key, val in count.items():
            bucket[val].append(key)

        # iterate through the bucket from back, return top k
        res = [num for b in reversed(bucket) for num in b]
        return res[:k]


        

